imports/wasi_snapshot_preview1/testdata/gotip/wasi.go

```diff
@@ -43,3 +51,52 @@ func mainSock() error {
 	fmt.Println(string(buf[:n]))
 	return nil
 }
+
+// mainHTTP implicitly tests non-blocking sockets, as they are needed for
+// middleware.
+func mainHTTP() error {
+	// Get the file representing a pre-opened TCP socket.
+	// The socket (listener) is the first pre-open, with a file-descriptor of
+	// 3 because the host didn't add any pre-opened files.
+	listenerFD := 3
+	f := os.NewFile(uintptr(listenerFD), "")
+
+	// Wasm runs similarly to GOMAXPROCS=1, so multiple goroutines cannot work
+	// in parallel. non-blocking allows the poller to park the go-routine
+	// accepting connections while work is done on one.
+	if err := syscall.SetNonblock(listenerFD, true); err != nil {
+		return err
+	}
+
+	// Convert the file representing the pre-opened socket to a listener, so
+	// that we can integrate it with HTTP middleware.
+	ln, err := net.FileListener(f)
+	defer f.Close()
+	if err != nil {
+		return err
+	}
+	defer ln.Close()
+
+	// Serve middleware that echos the request body to the response once, then quits.
+	h := &echoOnce{ch: make(chan struct{}, 1)}
+	go http.Serve(ln, h)
+	<-h.ch
+	return nil
+}
+
+type echoOnce struct {
+	ch chan struct{}
+}
+
+func (e echoOnce) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	// Copy up to 32 bytes from the request to the response, appending a newline.
+	// Note: the test should write: "wazero", so that's all we should read.
+	var buf [32]byte
+	if n, err := r.Body.Read(buf[:]); err != nil && err != io.EOF {
+		panic(err)
+	} else if n, err = w.Write(append(buf[:n], '\n')); err != nil {
+		panic(err)
+	}
+	// Once one request was served, close the channel.
+	close(e.ch)
```

Comment:
I think this is what I didn't get right when I tried to have the handler quit after one call and not loop back into the accept loop. glad you got it!